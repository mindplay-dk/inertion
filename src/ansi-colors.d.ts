// NOTE: copied from https://github.com/doowb/ansi-colors/blob/a4794363369d7b4d1872d248fc43a12761640d8e/types/index.d.ts
//
// TODO once this issue is resolved, switch back to `ansi-colors` proper
//      https://github.com/parcel-bundler/parcel/issues/7904

// Imported from DefinitelyTyped project.
// TypeScript definitions for ansi-colors
// Definitions by: Rogier Schouten <https://github.com/rogierschouten>
// Integrated by: Jordan Mele <https://github.com/Silic0nS0ldier>

type StyleArrayStructure = [number, number];
interface StyleArrayProperties {
  open: string;
  close: string;
  closeRe: string;
}

type StyleType = StyleArrayStructure & StyleArrayProperties;


interface StylesType<T> {
  // modifiers
  reset: T;
  bold: T;
  dim: T;
  italic: T;
  underline: T;
  inverse: T;
  hidden: T;
  strikethrough: T;

  // colors
  black: T;
  red: T;
  green: T;
  yellow: T;
  blue: T;
  magenta: T;
  cyan: T;
  white: T;
  gray: T;
  grey: T;

  // bright colors
  blackBright: T;
  redBright: T;
  greenBright: T;
  yellowBright: T;
  blueBright: T;
  magentaBright: T;
  cyanBright: T;
  whiteBright: T;

  // background colors
  bgBlack: T;
  bgRed: T;
  bgGreen: T;
  bgYellow: T;
  bgBlue: T;
  bgMagenta: T;
  bgCyan: T;
  bgWhite: T;

  // bright background colors
  bgBlackBright: T;
  bgRedBright: T;
  bgGreenBright: T;
  bgYellowBright: T;
  bgBlueBright: T;
  bgMagentaBright: T;
  bgCyanBright: T;
  bgWhiteBright: T;
}

declare namespace ansiColors {
  interface StyleFunction extends StylesType<StyleFunction> {
    (s: string): string;
  }

  // modifiers
  const reset: StyleFunction;
  const bold: StyleFunction;
  const dim: StyleFunction;
  const italic: StyleFunction;
  const underline: StyleFunction;
  const inverse: StyleFunction;
  const hidden: StyleFunction;
  const strikethrough: StyleFunction;

  // colors
  const black: StyleFunction;
  const red: StyleFunction;
  const green: StyleFunction;
  const yellow: StyleFunction;
  const blue: StyleFunction;
  const magenta: StyleFunction;
  const cyan: StyleFunction;
  const white: StyleFunction;
  const gray: StyleFunction;
  const grey: StyleFunction;

  // bright colors
  const blackBright: StyleFunction;
  const redBright: StyleFunction;
  const greenBright: StyleFunction;
  const yellowBright: StyleFunction;
  const blueBright: StyleFunction;
  const magentaBright: StyleFunction;
  const cyanBright: StyleFunction;
  const whiteBright: StyleFunction;

  // background colors
  const bgBlack: StyleFunction;
  const bgRed: StyleFunction;
  const bgGreen: StyleFunction;
  const bgYellow: StyleFunction;
  const bgBlue: StyleFunction;
  const bgMagenta: StyleFunction;
  const bgCyan: StyleFunction;
  const bgWhite: StyleFunction;

  // bright background colors
  const bgBlackBright: StyleFunction;
  const bgRedBright: StyleFunction;
  const bgGreenBright: StyleFunction;
  const bgYellowBright: StyleFunction;
  const bgBlueBright: StyleFunction;
  const bgMagentaBright: StyleFunction;
  const bgCyanBright: StyleFunction;
  const bgWhiteBright: StyleFunction;

  let enabled: boolean;
  let visible: boolean;
  const ansiRegex: RegExp;

  /**
   * Remove styles from string
   */
  function stripColor(s: string): string;

  /**
   * Remove styles from string
   */
  function strip(s: string): string;

  /**
   * Remove styles from string
   */
  function unstyle(s: string): string;

  const styles: StylesType<StyleType>;

  /**
   * Outputs a string with check-symbol as prefix
   */
  function ok(...args: string[]): string;

  function create(): typeof ansiColors;
}

export = ansiColors;
