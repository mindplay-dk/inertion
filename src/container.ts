export type Async<T> = {
  [K in keyof T as `$${string & K}`]: Promise<T[K]>;
}

export type FactoryMap<T> = {
  [K in keyof T]: (f: Omit<T, K>) => T[K];
}

export function createContainer<T>(factory: FactoryMap<T>): T {
  const container = {} as T;

  const instances = {} as { [key: string]: any };
  
  const has = Object.prototype.hasOwnProperty.bind(instances);

  for (const key in factory) {
    Object.defineProperty(
      container,
      key,
      {
        get() {
          if (! has(key)) {
            instances[key] = factory[key](container);
          }

          return instances[key];
        },
        configurable: true,
        enumerable: true,
      }
    );
  }

  return container;
}
