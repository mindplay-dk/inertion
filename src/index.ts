export * from "./api";
export * from "./harness";
export * from "./reporting";
export { default as assertions, assertion } from "./assertions";
