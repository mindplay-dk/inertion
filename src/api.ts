import { UnknownError } from "./harness";

export type MethodMap = {
  [name: string]: TestMethod;
}

export type TestMethod = {
  (...args: any[]): Fact;
}

export type ContextFactory<T> = () => T;

export type TestFactory<T extends MethodMap, C> = {
  (description: string, spec: Spec<T, C>): Test<T, C>;
}

export type Spec<T extends MethodMap, C> = {
  (tester: Tester<T>, context: C): Promise<void>;
}

export type Test<T extends MethodMap, C> = {
  methods: T;
  createContext: ContextFactory<C>;
  description: string;
  spec: Spec<T, C>;
}

export type Tester<T extends MethodMap> = {
  [K in keyof T]: (...args: Parameters<T[K]>) => void;
}

export type Fact = {
  label: string;
  pass: boolean;
  actual: unknown;
  expected: unknown;
  details: unknown[];
}

export type Check = {
  location: string;
  fact: Fact;
}

export type Result = {
  description: string;
  checks: Check[];
  time: number;
  error: Error | UnknownError | undefined;
}
