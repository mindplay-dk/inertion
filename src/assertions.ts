import isEqual from "fast-deep-equal";
import { Fact } from "./api";

export function ok(actual: unknown, ...details: unknown[]): Fact {
  return {
    label: "ok",
    pass: actual === true,
    actual,
    expected: true,
    details,
  };
}

// TODO IOC for `isEqual`: must support at least all of the following:
//      https://www.npmjs.com/package/fast-deep-equal    *DEFAULT*
//      https://www.npmjs.com/package/deep-equal
//      https://www.npmjs.com/package/deep-is

export function equal<T>(actual: T, expected: T, ...details: unknown[]): Fact {
  return {
    label: "equal",
    pass: isEqual(actual, expected),
    actual,
    expected,
    details,
  };
}

export default {
  ok,
  equal,
};

// TODO check compatibility of `assertion` with the following:
//      https://www.npmjs.com/package/validator
//      https://www.npmjs.com/package/check-types
//      https://www.npmjs.com/package/chai
//      https://www.npmjs.com/package/is
//      https://www.npmjs.com/package/typed-assert

/**
 * Generates a test-method from a simple assertion function.
 * 
 * This factory function assumes an input function where the first
 * argument is the actual value. Any additional parameters will be
 * reported as `details` in the generated `Fact` - the resulting
 * function will not have an `expected` value, and therefore will
 * be unable to display as a diff in reporters.
 * 
 * Works with predicate libraries such as `validator` or `is`.
 * 
 * @see https://www.npmjs.com/package/validator
 */
export function assertion<F extends (...args: any[]) => boolean>(label: string, assert: F) {
  return (...args: [...params: Parameters<F>, ...details: unknown[]]): Fact => ({
    label,
    pass: assert(...args.slice(0, assert.length)),
    actual: args[0],
    expected: undefined,
    details: args.slice(1),
  });
}
