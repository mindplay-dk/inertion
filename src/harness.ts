import { Check, MethodMap, Result, Test, Tester, TestFactory } from "./api";

export function setup<T extends MethodMap>(methods: T): TestFactory<T> {
  return (description, spec) => ({ methods, description, spec });
}

export async function run(tests: Array<Test<any>>): Promise<Result[]> {
  return Promise.all(tests.map(runTest));
}

export async function runTest<T extends MethodMap>({ methods, description, spec }: Test<T>): Promise<Result> {
  const checks: Check[] = [];

  let checked = (check: Check) => {
    checks.push(check);
  };

  const tester = Object.fromEntries(
    Object.entries(methods)
      .map(([name, assertion]) => [name, (...args: any) => {
        const location = new Error().stack!
          .split("\n")[2]
          .replace(/^\s+at /, "");

        const check = assertion(...args);

        checked({ location, fact: check });
      }])
  ) as Tester<T>;

  const started = Date.now();

  let error: unknown;

  try {
    await spec(tester);
  } catch (e: unknown) {
    error = e;
  }

  const time = Date.now() - started;

  checked = () => {
    throw new Error(`attempted assertion after end of test "${description}" - please check your code for missing await statements.`);
  };

  return {
    description,
    checks,
    time,
    error,
  };
}
