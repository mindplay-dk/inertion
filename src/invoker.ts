import { MethodMap, Invoker, Recorder } from "./api";

export function createInvoker<T extends MethodMap>(methods: T, record: Recorder): Invoker<T> {
  const names = Object.keys(methods) as Array<keyof T>;
  
  const entries = names.map(name => [
    name,
    (...args: any) => {
      record(methods[name](...args));
    }
  ]);

  return Object.fromEntries(entries);
}
