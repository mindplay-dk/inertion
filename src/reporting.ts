import { Result } from "./api";
import util from "util";

export function statusOf(results: Result[]): number {
  return isSuccess(results) ? 1 : 0;
}

export function isSuccess(results: Result[]): boolean {
  return results.some(isFailed) === false;
}

export function isFailed(result: Result): boolean {
  return !!result.error || result.checks.some(({ fact }) => !fact.pass);
}

const inspect = (...values: unknown[]) => util.inspect(values, { depth: 99, maxArrayLength: 999, colors: true });

export function reportTo(console: Console, results: Result[]): void {
  // TODO add proper reporting
  //
  //      - pick a formatter:
  //        https://github.com/facebook/jest/tree/main/packages/pretty-format
  //        https://github.com/hildjj/node-inspect-extracted
  //
  //      - or fork this one?
  //        https://github.com/commenthol/serialize-to-js
  //
  //      - add diffing:
  //        https://www.npmjs.com/package/diff
  //        
  //        see note here about picking between character/line diff:
  //        https://github.com/lorenzofox3/zora/issues/127

  console.log(`${results.length} tests completed\n`);

  for (const result of results) {
    console.log(`"${result.description}"\n`);

    for (const { fact: check, location } of result.checks) {
      console.log(
        check.pass ? "✅" : "❌",
        `[${check.label}]`,
        ...check.details
      );

      if (!check.pass) {
        console.log(`  location:`, location);
        console.log(`  actual:  `, inspect(check.actual));
        console.log(`  expected:`, inspect(check.expected));
      }
    }
  }
}
