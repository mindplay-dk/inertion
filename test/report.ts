import { Check, Fact, Result } from "../src/api";
import { UnknownError } from "../src/harness";
import { isFailed } from "../src/reporting";

interface Report {
  passed: number;
  failed: number;
  time: number;
  results: Array<{
    description: string;
    pass: boolean;
    time: number;
    error: string | undefined;
    checksPassed: number;
    checksFailed: number;
    facts: Array<{
      label: string;
      pass: boolean;
      actual: string | undefined;
      expected: string | undefined;
      details: string | undefined;
    }>;
  }>;
}

type Formatter = (value: unknown) => string;

function createReport(results: Result[], format: Formatter): Report {
  const failed = results.filter(isFailed).length;
  const passed = results.length - failed;
  
  let time = 0;

  for (const result of results) {
    time += result.time;
  }

  return {
    passed,
    failed,
    time,
    results: results.map(result => ({
      description: result.description,
      pass: ! isFailed(result),
      time: result.time,
      error: result.error
        ? (result.error instanceof UnknownError ? `Unknown error:\n` + format(result.error.value) : "") + result.error.stack
        : undefined,
      checksPassed: result.checks.filter(({ fact }) => fact.pass).length,
      checksFailed: result.checks.filter(({ fact }) => ! fact.pass).length,
      facts: result.checks.map(({ fact: { label, pass, actual, expected, details }, location }) => ({
        label,
        pass,
        actual: actual !== undefined ? format(actual) : undefined,
        expected: expected !== undefined ? format(expected) : undefined,
        details: typeof details[0] === "string"
          ? details[0] + (details.length > 1 ? "\n" + format(details.slice(1)) : "")
          : format(details)
      })),
    })),
  };
}

// verbosity options:
//
//   0: print descriptions of failed tests only
//   1: print all descriptions (with check counts)
//   2: print all descriptions (with check counts) and checks (with details)

type Verbosity = 0 | 1 | 2;

const verbosityOptions: Verbosity[] = [0, 1, 2];

function createSampleResults(): Result[] {
  let results: Result[] = [];

  // we want to see all combinations of actual/expected values and pass/fail:

  const valueModes: Record<string, Pick<Fact, "actual" | "expected">> = {
    "no actual or expected": { actual: undefined, expected: undefined },
    "actual only (single-line)": { actual: "actual only, single-line", expected: undefined },
    "actual only (multi-line)": { actual: "actual only\nmulti-line", expected: undefined },
    "actual and expected (both single-line)": { actual: "actual, single-line", expected: "expected, single-line" },
    "actual and expected (multi-line)": { actual: "actual\nmulti-line", expected: "expected\nmulti-line" },
  };

  for (const [values, valueMode] of Object.entries(valueModes)) {
    let checks: Check[] = [];

    for (const pass of [true, false]) {
      checks.push({
        location: `at test.js:1:1`,
        fact: {
          label: "ok",
          pass,
          ...valueMode,
          details: [],
        }
      });
    }

    results.push({ description: values, time: 0.123, error: undefined, checks });
  }

  // we want to see some basic pass/fail checks in combination with different error modes:

  const basicChecks: Check[] = [true, false].map(pass => ({
    location: `at test.js:1:1`,
    fact: {
      label: "ok",
      pass,
      actual: undefined,
      expected: undefined,
      details: []
    }
  }));

  const resultModes: Record<string, Pick<Result, "error" | "checks">> = {
    "no error": { error: undefined, checks: basicChecks },
    "has error": { error: new Error(`oh no`), checks: basicChecks },
    "unknown error": { error: new UnknownError({ oops: "oh no" }), checks: basicChecks },
  };
 
  for (const [result, resultMode] of Object.entries(resultModes)) {
    results.push({
      description: result,
      ...resultMode,
      time: 0.123,
    });
  }

  // we want to see some basic pass/fail checks in combination with various modes of detail:

  const detailModes: Record<string, Pick<Fact, "details">> = {
    "no details": { details: [] },
    "one single-line detail": { details: ["details, single-line"] },
    "multiple lines or details": { details: ["details\nmulti-line", "more details"] },
  };

  for (const [details, detailMode] of Object.entries(detailModes)) {
    results.push({
      description: details,
      error: undefined,
      checks: basicChecks.map(check => ({ ...check, ...detailMode })),
      time: 0.123,
    });
  }

  return results;
}

console.log(JSON.stringify(createReport(createSampleResults(), v => JSON.stringify(v, null, 2))));
