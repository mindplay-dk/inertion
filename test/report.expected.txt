FAILED: no actual or expected (0.123ms)

  Ã— [ok] no actual or expected
  â”” test.js:1:1

FAILED: actual only (single-line) (0.123ms)

  Ã— [ok] actual only (single-line)
  â”” test.js:1:1

  [41m[37m Ã— [39m[49m ACTUAL: "actual only, single-line"

FAILED: actual only (multi-line) (0.123ms)

  Ã— [ok] actual only (multi-line)
  â”” test.js:1:1

  [41m[37m Ã— [39m[49m ACTUAL:
      "actual only\n"
      + "multi-line"

FAILED: actual and expected (single-line, same types) (0.123ms)

  Ã— [ok] actual and expected (single-line, same types)
  â”” test.js:1:1

  ACTUAL:   "[41m[37mactual[39m[49m, single-line"
  EXPECTED: "[42m[37mexpected[39m[49m, single-line"

FAILED: actual and expected (single-line, different types) (0.123ms)

  Ã— [ok] actual and expected (single-line, different types)
  â”” test.js:1:1

  [41m[37m Ã— [39m[49m ACTUAL:   "actual"
  [42m[37m âˆš [39m[49m EXPECTED: 123

FAILED: actual and expected (multi-line, same types) (0.123ms)

  Ã— [ok] actual and expected (multi-line, same types)
  â”” test.js:1:1

      Object {
        "same": true,
  [41m[37m Ã— [39m[49m   "value": "actual\n"
  [42m[37m âˆš [39m[49m   "value": "expected\n"
        + "multi-line",
      }

FAILED: actual and expected (multi-line, different types) (0.123ms)

  Ã— [ok] actual and expected (multi-line, different types)
  â”” test.js:1:1

  [41m[37m Ã— [39m[49m ACTUAL:   "actual"
  [42m[37m âˆš [39m[49m EXPECTED: Foo { ... }

FAILED: actual and expected (multi-line, additions) (0.123ms)

  Ã— [ok] actual and expected (multi-line, additions)
  â”” test.js:1:1

      Array [
        1,
        2,
  [42m[37m âˆš [39m[49m   3,
      ]

FAILED: actual and expected (multi-line, removals) (0.123ms)

  Ã— [ok] actual and expected (multi-line, removals)
  â”” test.js:1:1

      Array [
        1,
        2,
  [41m[37m Ã— [39m[49m   3,
      ]

FAILED: no error (0.123ms)

  Ã— [ok] 
  â”” test.js:1:1

FAILED: has error (0.123ms)
  Error: oh no
      at createSampleResults (/path/to/test/report.ts:77:27)
      at import_reporter (/path/to/test/report.ts:128:13)
      at Object.<anonymous> (/path/to/test/report.ts:128:34)
      at Object.j (/path/to/node_modules/tsx/dist/cjs/index.cjs:1:1197)

  Ã— [ok] 
  â”” test.js:1:1

FAILED: unknown error (0.123ms)
  Unknown error:
  Object {
    "oops": "oh no",
  }
  Error: Unknown error type
      at createSampleResults (/path/to/test/report.ts:78:31)
      at import_reporter (/path/to/test/report.ts:128:13)
      at Object.<anonymous> (/path/to/test/report.ts:128:34)
      at Object.j (/path/to/node_modules/tsx/dist/cjs/index.cjs:1:1197)

  Ã— [ok] 
  â”” test.js:1:1

FAILED: no details (0.123ms)

  Ã— [ok] 
  â”” test.js:1:1

FAILED: one single-line detail (0.123ms)

  Ã— [ok] details, single-line
  â”” test.js:1:1

FAILED: multiple lines or details (0.123ms)

  Ã— [ok] 
  â”” test.js:1:1
  DETAILS:
  Array [
    "details\n"
    + "multi-line",
    "more details",
  ]

âˆš PASSED: 0
Ã— FAILED: 15
  TOTAL:  15
